code/modules/reagents/chemistry/machinery/chem_master.dm

```diff
@@ -425,21 +425,24 @@
 
 		if("transfer")
 			if(is_printing)
-				say("buffer locked while printing!")
+				say("The buffer is locked while printing.")
 				return
 
 			var/reagent_ref = params["reagentRef"]
 			var/amount = params["amount"]
 			var/target = params["target"]
-
+			var/should_transfer = transfer_mode || (target == "buffer") // we should always transfer if target is the buffer
+			if(should_transfer && isnull(beaker)) // if there's no beaker, we cannot transfer
+				say("No reagent container is inserted.")
+				return FALSE
 			if(target == "buffer")
-				return transfer_reagent(ui.user, beaker.reagents, reagents, reagent_ref, amount, TRUE)
+				return transfer_reagent(ui.user, beaker?.reagents, reagents, reagent_ref, amount, should_transfer)
 			else if(target == "beaker")
-				return transfer_reagent(ui.user, reagents, beaker.reagents, reagent_ref, amount, is_transfering)
+				return transfer_reagent(ui.user, reagents, beaker?.reagents, reagent_ref, amount, should_transfer)
```

Comment:
this code seems really weird to me. I don't like the ?.reagents bit since it just silences a potential runtime, but then looking at it if should_transfer is FALSE it's valid?

transfer_reagent as a proc is doing too much shit. If we're trying to modularize we should split it up into code that asks for an amount, consumes power, and then use those procs to rebuild behavior here. at least IMO. this is maybe not the best solution but having args that only sometimes do anything is one hell of a code smell. I'd like to fix the structure here rather then just a symptom of the structure.

Reply:
yea I figured that you don't actually need `beaker.reagents` when the task is to delete reagents from the buffer. I wanted to make a separate proc for that case, but then I looked at `transfer_reagent` itself and found out that most of it is just sanity checks & the actual functionality is like 3 lines of code which call some other procs, so my options for changing the structure were 
1) move all the sanity checks to `ui_act` which seemed bad,
2) separate them into another proc (which would practically lead to the same structure as before, just more bloated) or 
3) make 2 procs with 80% of them being the same checks, which smelled just as bad.
In the end, I figured this is the better solution, maybe I'm missing something but I just don't know how else I can change the structure of this thing

Reply:
that said I think I can at least clean up `transfer_reagent` a little